pkg/controller/eventtype/reconcile.go

```diff
@@ -0,0 +1,215 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventtype
+
+import (
+	"context"
+	"fmt"
+	"github.com/golang/glog"
+	feedsv1alpha1 "github.com/knative/eventing/pkg/apis/feeds/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+)
+
+const (
+	finalizerName = controllerAgentName
+)
+
+// Reconcile compares the actual state of a Feed with the desired, and attempts
+// to converge the two. It then updates the Status block of the Feed with
+// its current state.
+// If Reconcile returns a non-nil error, the request will be retried.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	//TODO use this to store the logger and set a deadline
+	ctx := context.TODO()
+
+	et := &feedsv1alpha1.EventType{}
+	err := r.client.Get(ctx, request.NamespacedName, et)
+
+	// The EventType may have been deleted since it was added to the workqueue. If so
+	// there's nothing to be done.
+	if errors.IsNotFound(err) {
+		glog.Errorf("could not find EventType %v\n", request)
+		return reconcile.Result{}, nil
+	}
+
+	// If the EventType exists but could not be retrieved, then we should retry.
+	if err != nil {
+		glog.Errorf("could not fetch EventType %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	// Now that we know the EventType exists, we can reconcile it. An error returned
+	// here means the reconcile did not complete and the EventType should be requeued
+	// for another attempt.
+	// A successful reconcile does not necessarily mean the EventType is in the desired
+	// state, it means no more can be done for now. In this case the EventType will
+	// not be reconciled again until the resync period or a watched resource changes.
+	if err = r.reconcile(ctx, et); err != nil {
+		glog.Errorf("error reconciling EventType: %v", err)
+	}
+
+	// Since the reconcile is a sequence of steps, earlier steps may complete
+	// successfully while later steps fail. The EventType is updated on failure to
+	// preserve any useful status or metadata changes the non-failing steps made.
+	if updateErr := r.updateEventType(ctx, et); updateErr != nil {
+		glog.Errorf("failed to update EventType: %v", updateErr)
+		// An error here means the EventType should be reconciled again, regardless of
+		// whether the reconcile was successful or not.
+		return reconcile.Result{}, updateErr
+	}
+	return reconcile.Result{}, err
```

Comment:
this err is from `err := r.client.Get(ctx, request.NamespacedName, et)`

is that intended? I think it is not, because err should be nil here, otherwise we would have returned above.

Reply:
I had the same question when first seeing this pattern in the [Feed controller](https://github.com/knative/eventing/blob/master/pkg/controller/feed/reconcile.go#L75).

Logically, what is happening, is that we first attempt to reconcile, never writing any updates to the API server (this is `r.reconcile()` on line 68). Then, after that is done, we attempt to write all updates to the API server (`r.updateEventType()` on line 75). 

If there are problems reconciling, then `r.reconcile()` should have updated the in-memory status of the EventType, but needs `r.updateEventType()` to actually write it to the API server. So we need to call `r.updateEventType()`, regardless of whether `r.reconcile()` returned an error. Then, after updating the API server, we need to know whether or not this reconcile was successful, so we return the error (which may or may not be nil).

I added a comment to the code where you would expect the original err to be returned, does that make it clearer what is going on?